accepted/2021/preview-features/preview-features.md

```diff
@@ -0,0 +1,418 @@
+
+### Can we make breaking changes in preview features?
+
+Yes! That's the whole point of marking features as preview in the first place.
+Please note that breaking change also includes yanking the preview feature in
+its entirety. Preview features should be treated like previews of .NET: we don't
+make promises and reserve the right to make changes.
+
+Halfway through a release cycle we might also decide that certain features won't
+make the cut for RTM quality and we might decide to mark features we shipped in
+a previous preview as `[RequiresPreviewFeatures]`. The reverse is also possible
+(and argubly not a breaking change) but probably less likely.
+
+### Why a single on/off switch as opposed to individual feature switches?
+
+I originally designed it as such:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ExperimentalFeatures>FancyGC;StaticInterfaceMembers</ExperimentalFeatures>
+  </PropertyGroup>
+
+</Project>
+```
+
+The upside is that one can be explicit about which preview feature one wants to
+turn on (and thus depend on).
+
+The downsides were:
+
+1. It's a more complicated concept for the user to understand and very different
+   from how previews of the .NET platform itself work. This might increase the
+   barrier of entry, which we don't want.
+
+2. The test matrix for the product gets much more complicated as the combinations
+   of on/states explode quickly.
+
+3. Not all preview features have a convenient entry point (language syntax,
+   API). Having a global preview mode increases airtime for features that
+   otherwise wouldn't have been turned on.
+
+4. We're unlikely to ship a grab bag of preview features and instead are more
+   likely to ship a cohesive feature set that in its totality is still in
+   preview, such as static interface members, which includes runtime work,
+   language work, and compiler work. Said differently, having a single preview
+   mode forces us to jointly coordinate preview features across all layers.
+
+### Why are we recording whether an assembly used preview features?
+
+For two reasons:
+
+1. It can be used for auditing
+2. It is used to enforce transitivity. An assembly that doesn't use preview
+   features shouldn't depend on assemblies that do. It doesn't matter whether
+   this binary was a project-to-project reference, a NuGet package, or a raw
+   reference to some checked in file -- we only have to look at the assembly
+   level attribute.
+
+### What happens when a preview feature is released?
+
+Several things will happen:
+
+1. We remove the `[RequiresPreviewFeatures]` attribution from any APIs
+2. We remove any guards from the runtime/compilers that check for the preview
+   flag (of course, compilers are likely still checking language versions, but
+   that's separate)
+
+If someone shipped a binary to nuget.org that depends on a preview feature,
```

Comment:
Should it be an error to attempt to package a 'stable' NuGet package containing assemblies with the preview attribute?

Reply:
I thought about it but I don't see a need for that if we track this via assembly level attributes.